Draft: Better way to resolve allOf

SOHELAHMED7 opened the pull request:
Fixes https://github.com/cebe/yii2-openapi/issues/165

Also part of https://github.com/php-openapi/yii2-openapi/issues/10

SOHELAHMED7 commented:
### Requirements:

Simple example of resolving a property:

Schema:

```yaml
openapi: 3.0.3

info:
  title: Resolve property
  version: 1.0.0

components:
  schemas:
    User:
      type: object
      required:
        - id
        - name
      properties:
        id:
          type: integer
        name:
          type: string
    Post:
      type: object
      properties:
        id:
          type: integer
        content:
          type: string
        user:
          $ref: '#/components/schemas/User'

paths:
  '/':
    get:
      responses:
        '200':
          description: OK
```

After resolving the references:

```yaml
components:
  schemas:
    User:
      type: object
      required:
        - id
        - name
      properties:
        id:
          type: integer
        name:
          type: string
    Post:
      type: object
      properties:
        id:
          type: integer
        content:
          type: string
        user:
          type: object
          required:
            - id
            - name
          properties:
            id:
              type: integer
            name:
              type: string
```


In the same manner `allOf` with reference and custom OpenAPI extension with the schema:

```yaml
    Post:
      type: object
      properties:
        id:
          type: integer
        content:
          type: string
        user:
          allOf:
            - $ref: '#/components/schemas/User'
            - x-faker: false
```


must be resolved as:

```yaml
    Post:
      type: object
      properties:
        id:
          type: integer
        content:
          type: string
        user:
          type: object
          required:
            - id
            - name
          properties:
            id:
              type: integer
            name:
              type: string
          x-faker: false
```

SOHELAHMED7 commented:
If a duplicate property is found

```yaml
components:
  schemas:
    User:
      type: object
      required:
        - id
        - name # <--------------------------------------------------------------
      properties:
        id:
          type: integer
        name: # <--------------------------------------------------------------
          type: string
          maxLength: 10 # <--------------------------------------------------------------
    Pet:
      type: object
      required:
        - id2
        - name # <--------------------------------------------------------------
      properties:
        id2:
          type: integer
        name: # <--------------------------------------------------------------
          type: string
          maxLength: 12 # <--------------------------------------------------------------
    Post:
      type: object
      properties:
        id:
          type: integer
        content:
          type: string
        user:
          allOf:
            - $ref: '#/components/schemas/User'
            - $ref: '#/components/schemas/Pet'
            - x-faker: true
```

then property from the last component schema will be considered:

```yaml
Post:
  type: object
  properties:
    id:
      type: integer
    content:
      type: string
    user:
      type: object
      required:
        - id
        - name # <--------------------------------------------------------------
        - id2
      properties:
        id:
          type: integer
        name: # <--------------------------------------------------------------
          type: string
          maxLength: 12 # <--------------------------------------------------------------
        id2:
          type: integer
      x-faker: true
```

simPod commented:
You'll have better chance in https://github.com/DEVizzent/cebe-php-openapi since this repo is abandoned.

cebe commented on src/spec/OpenApi.php:
```diff
@@ -79,4 +79,47 @@ public function performValidation()
+     * Function call example: `$result = array_merge_recursive_distinct(a1, a2, ... aN);`
+     * More documentation is present at [array merge recursive distinct.md](../../../doc/array merge recursive distinct.md) file
+     */
+    public static function arrayMergeRecursiveDistinct()
```
this should be a method of a separate helper class.

SOHELAHMED7 replied:
Implemented in 81ee4edb6882943f846d95f37f3d26e42eb66662

cebe reviewed (commented):
Not sure about the recursive call. This can be a significant performance problem. I'd keep the resolveAllOf part on the Schema only and call it if needed.

cebe commented:
@SOHELAHMED7 can you explain more about why you implemented the recursive resolution of allOf? allOf is a property of Schema and it only make sense to call it on a schema object. I'd expect the resolveAllOf() function inside of Schema and see it applied to sub-schemas only.

SOHELAHMED7 commented:
> can you explain more about why you implemented the recursive resolution of allOf?

I think `allOf` and "reference" are very similar. "reference" can be nested so as `allOf`. `allOf` can have "reference".  [I have added a test case for this.](0bc15f5d1920ca21ae59851b1b54918832df0105). Also resolving `allOf` will require resolving "reference".

So I implemented resolving `allOf` similar to how "reference" are resolved.

cebe commented on src/SpecBaseObject.php:
```diff
@@ -525,4 +527,86 @@ public function getExtensions(): array
+        }
+    }
+
+    public function removeAllOfKey(string $property, $value): void
```
```suggestion
    private function removeAllOfKey(string $property, $value): void
```

cebe commented on src/SpecBaseObject.php:
```diff
@@ -525,4 +527,86 @@ public function getExtensions(): array
+        return $first;
+    }
+
+    public function handleMergingOfAllAllOfs(string $property, $value): void
```
```suggestion
    private function handleMergingOfAllAllOfs(string $property, $value): void
```

cebe commented on src/SpecBaseObject.php:
```diff
@@ -525,4 +527,86 @@ public function getExtensions(): array
+        $this->_recursingAllOf = false;
+    }
+
+    public function mergeAllAllOfsInToSingleObject(): self
```
```suggestion
    private function mergeAllAllOfsInToSingleObject(): self
```
